internal/pkg/db/redis/data.go

```diff
@@ -203,6 +217,122 @@ func (c *Client) DeleteEventById(id string) (err error) {
+	}
+
+	conn := c.Pool.Get()
+	defer conn.Close()
+
+	ids, err := redis.Strings(conn.Do("ZRANGE", db.EventsCollection+":device:"+deviceId, 0, -1))
+	if err != nil {
+		return 0, err
+	}
+
+	err = conn.Send("MULTI")
+	if err != nil {
+		return 0, err
+	}
+
+	for _, id := range ids {
+		err = conn.Send("RENAME", id, db.DeletedEventsCollection+":"+id)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	err = conn.Send("EXEC")
+	deleteEventsChannel <- deviceId
+
+	return len(ids), nil
+}
+
+// AsyncDeleteEvents Handles the deletion of device events asynchronously. This function is expected to be running in
+// a go-routine and works with the "DeleteEventsByDevice" function for better performance.
+func (c *Client) AsyncDeleteEvents(loggingClient logger.LoggingClient) {
+	loggingClient.Debug("Starting background event deletion process")
+	for {
+		select {
+		case device, ok := <-deleteEventsChannel:
+			if ok {
+				loggingClient.Debug("Deleting event data for device: " + device)
+				startTime := time.Now()
+				c.deleteRenamedEvents(device, loggingClient)
+				loggingClient.Debug(fmt.Sprintf("Deleted events for device: '%s', elapsed time: %s", device, time.Since(startTime)))
+			}
+		}
+	}
+}
+
+// deleteRenamedEvents deletes all events associated with the specified device which have been marked for deletion.
+// See the "DeleteEventsByDevice" function for details on the how events are marked for deletion(renamed)
+func (c *Client) deleteRenamedEvents(device string, loggingClient logger.LoggingClient) {
+	conn := c.Pool.Get()
+	defer conn.Close()
+
+	ids, err := redis.Strings(conn.Do("ZRANGE", db.EventsCollection+":device:"+device, 0, -1))
+	if err != nil {
+		loggingClient.Error("Unable to delete event:" + err.Error())
+		return
+	}
+
+	_, err = conn.Do("MULTI")
+	if err != nil {
+		loggingClient.Error("Unable to start transaction for deletion:" + err.Error())
+	}
+
+	for _, id := range ids {
+		_, err = conn.Do("GET", db.DeletedEventsCollection+":"+id)
+		if err != nil {
+			loggingClient.Error("Unable to obtain events marked for deletion:" + err.Error())
+		}
+	}
+	events, err := redis.Strings(conn.Do("EXEC"))
+
+	queriesInQueue := 0
+	var e correlation.Event
+	_, err = conn.Do("MULTI")
+	if err != nil {
+		loggingClient.Error("Unable to start batch processing for event deletion:" + err.Error())
+	}
+
+	for _, event := range events {
+		err = unmarshalObject([]byte(event), &e)
+		if err != nil {
+			loggingClient.Error("Unable to marshal event: " + err.Error())
+		}
+		_ = conn.Send("UNLINK", db.DeletedEventsCollection+":"+e.ID)
+		_ = conn.Send("ZREM", db.EventsCollection, e.ID)
+		_ = conn.Send("ZREM", db.EventsCollection+":created", e.ID)
+		_ = conn.Send("ZREM", db.EventsCollection+":device:"+e.Device, e.ID)
+		_ = conn.Send("ZREM", db.EventsCollection+":pushed", e.ID)
+		if e.Checksum != "" {
+			_ = conn.Send("ZREM", db.EventsCollection+":checksum:"+e.Checksum, 0)
+		}
+
+		queriesInQueue++
+		if queriesInQueue >= AsyncDeleteBatchSize {
```

Comment:
Suggest `c. AsyncDeleteBatchSize` driven from configuration as mentioned in comment above -- tune-able as performance needs / resources warrant.